File: CSharp/SpaceVIL/SpaceVIL/Items/ListBox.cs
using System;
using System.Linq;
using System.Drawing;
using System.Collections.Generic;
using SpaceVIL.Core;
using SpaceVIL.Decorations;
using SpaceVIL.Common;

namespace SpaceVIL
{
    public class ListBox : Prototype
    {
        static int count = 0;

        /// <summary>
        /// ScrollBar moving step
        /// </summary>
        public void SetScrollStep(int step)
        {
            _area.SetStep(step);
        }
        public int GetScrollStep()
        {
            return _area.GetStep();
        }

        /// <summary>
        /// Selection item
        /// </summary>
        public int GetSelection()
        {
            return _area.GetSelection();
        }

        /// <summary>
        /// Set selected item of the ListBox by index
        /// </summary>
        public void SetSelection(int index)
        {
            _area.SetSelection(index);
        }

        /// <summary>
        /// Unselect all items
        /// </summary>
        public void Unselect()
        {
            _area.Unselect();
        }

        /// <summary>
        /// Is selection changes view of the item or not
        /// </summary>
        public void SetSelectionVisible(bool value)
        {
            _area.SetSelectionVisible(value);
        }
        public bool IsSelectionVisible()
        {
            return _area.IsSelectionVisible();
        }

        public BlankItem Menu = new BlankItem();
        private bool _is_menu_disabled = false;

        /// <summary>
        /// Is ListBox menu disabled
        /// </summary>
        public void DisableMenu(bool value)
        {
            _is_menu_disabled = value;
        }

        private Grid _grid = new Grid(2, 2);
        private ListArea _area = new ListArea();

        /// <returns> ListArea </returns>
        public ListArea GetArea()
        {
            return _area;
        }

        private ContextMenu _menu;
        public VerticalScrollBar VScrollBar = new VerticalScrollBar();
        public HorizontalScrollBar HScrollBar = new HorizontalScrollBar();
        private ScrollBarVisibility _v_scrollBarPolicy = ScrollBarVisibility.AsNeeded;

        /// <summary>
        /// Is vertical scroll bar visible
        /// </summary>
        public ScrollBarVisibility GetVScrollBarVisible()
        {
            return _v_scrollBarPolicy;
        }
        public void SetVScrollBarVisible(ScrollBarVisibility policy)
        {
            _v_scrollBarPolicy = policy;

            if (policy == ScrollBarVisibility.Never)
            {
                VScrollBar.SetDrawable(false);
                Menu.SetVisible(false);
            }
            else if (policy == ScrollBarVisibility.AsNeeded)
            {
                VScrollBar.SetDrawable(false);
                Menu.SetVisible(false);
            }
            else if (policy == ScrollBarVisibility.Always)
            {
                VScrollBar.SetDrawable(true);
                if (!HScrollBar.IsDrawable())
                    Menu.SetVisible(false);
                else
                    Menu.SetVisible(true);
            }

            _grid.UpdateLayout();
            UpdateHorizontalSlider();
            HScrollBar.Slider.UpdateHandler();
        }

        private ScrollBarVisibility _h_scrollBarPolicy = ScrollBarVisibility.AsNeeded;

        /// <summary>
        /// Is horizontal scroll bar visible
        /// </summary>
        public ScrollBarVisibility GetHScrollBarVisible()
        {
            return _h_scrollBarPolicy;
        }
        public void SetHScrollBarVisible(ScrollBarVisibility policy)
        {
            _h_scrollBarPolicy = policy;

            if (policy == ScrollBarVisibility.Never)
            {
                HScrollBar.SetDrawable(false);
                Menu.SetVisible(false);
            }
            else if (policy == ScrollBarVisibility.AsNeeded)
            {
                HScrollBar.SetDrawable(false);
                Menu.SetVisible(false);
            }
            else if (policy == ScrollBarVisibility.Always)
            {
                HScrollBar.SetDrawable(true);
                if (!VScrollBar.IsDrawable())
                    Menu.SetVisible(false);
                else
                    Menu.SetVisible(true);
            }

            _grid.UpdateLayout();
            UpdateVerticalSlider();
            VScrollBar.Slider.UpdateHandler();
        }

        /// <summary>
        /// Constructs a ListBox
        /// </summary>
        public ListBox()
        {
            SetItemName("ListBox_" + count);
            count++;
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ListBox)));

            //VBar
            VScrollBar.SetDrawable(true);
            VScrollBar.SetItemName(GetItemName() + "_" + VScrollBar.GetItemName());

            //HBar
            HScrollBar.SetDrawable(true);
            HScrollBar.SetItemName(GetItemName() + "_" + HScrollBar.GetItemName());

            //events
            EventMouseClick += (sender, args) =>
            {
                // зачем нужен этот if???   
                if (_area.GetSelection() >= 0)
                    _area.SetSelection(_area.GetSelection());
                _area.SetFocus();
            };
            EventKeyPress += OnKeyPress;
        }

        private bool IsOutsideArea(long y, int h, Indents margin)
        {
            int startY = _area.GetY() + GetPadding().Top;
            int endY = _area.GetY() + GetHeight() - GetPadding().Bottom;
            if (y + h < startY || y - margin.Top > endY)
                return true;
            return false;
        }

        private void OnKeyPress(IItem sender, KeyArgs args)
        {
            SelectionItem selection = _area.GetTrueSelection();
            if (selection == null)
                return;
            long offset = _area.GetVScrollOffset();
            int startY = _area.GetY() + GetPadding().Top;
            long selection_Y = selection.GetY() + selection.GetMargin().Top;

            switch (args.Key)
            {
                case KeyCode.Up:
                    if (IsOutsideArea(selection_Y, selection.GetHeight(), selection.GetMargin()))
                    {
                        _area.SetVScrollOffset(offset - (startY - selection_Y));
                        UpdateVerticalSlider();
                        break;
                    }
                    if (selection_Y < startY)
                    {
                        _area.SetVScrollOffset(offset - (startY - selection_Y));
                        UpdateVerticalSlider();
                    }
                    break;
                case KeyCode.Down:
                    if (IsOutsideArea(selection_Y, selection.GetHeight(), selection.GetMargin()))
                    {
                        // _area.SetVScrollOffset(offset - (startY - selection_Y));
                        _area.SetVScrollOffset(offset + selection.GetHeight() + (selection.GetY() - (_area.GetY() + _area.GetHeight())));
                        UpdateVerticalSlider();
                        break;
                    }
                    if (selection_Y + selection.GetHeight() + selection.GetMargin().Bottom > GetY() + GetHeight() - GetPadding().Bottom)
                    {
                        _area.SetVScrollOffset(offset + ((selection_Y + selection.GetHeight() + selection.GetMargin().Bottom
                                + _area.GetSpacing().Vertical) - (GetY() + GetHeight() - GetPadding().Bottom)));
                        UpdateVerticalSlider();
                    }
                    break;
                default:
                    break;
            }
        }

        private void UpdateListAreaAttributes(object sender)
        {
            UpdateVListArea();
            UpdateHListArea();
        }

        private Int64 v_size = 0;
        private Int64 h_size = 0;

        private void UpdateVListArea()
        {
            //vertical slider
            float v_value = VScrollBar.Slider.GetCurrentValue();
            int v_offset = (int)Math.Round((float)(v_size * v_value) / 100.0f);
            _area.SetVScrollOffset(v_offset);
        }
        private void UpdateHListArea()
        {
            //horizontal slider
            float h_value = HScrollBar.Slider.GetCurrentValue();
            int h_offset = (int)Math.Round((float)(h_size * h_value) / 100.0f);
            _area.SetHScrollOffset(h_offset);
        }

        private void UpdateVerticalSlider()//vertical slider
        {
            int total_invisible_size = 0;
            int visible_area = _area.GetHeight() - _area.GetPadding().Top - _area.GetPadding().Bottom;
            if (visible_area < 0)
                visible_area = 0;
            foreach (var item in _area.GetItems())
            {
                if (!item.IsVisible())
                    continue;
                total_invisible_size += (item.GetHeight() + _area.GetSpacing().Vertical);
            }
            int total = total_invisible_size - _area.GetSpacing().Vertical;
            if (total_invisible_size <= visible_area)
            {
                VScrollBar.Slider.Handler.SetDrawable(false);
                VScrollBar.Slider.SetStep(VScrollBar.Slider.GetMaxValue());
                v_size = 0;
                VScrollBar.Slider.SetCurrentValue(0);
                if (GetVScrollBarVisible() == ScrollBarVisibility.AsNeeded)
                {
                    VScrollBar.SetDrawable(false);
                    Menu.SetVisible(false);
                    _grid.UpdateLayout();
                }
                return;
            }
            if (GetVScrollBarVisible() == ScrollBarVisibility.AsNeeded)
            {
                VScrollBar.SetDrawable(true);
                if (!HScrollBar.IsDrawable())
                    Menu.SetVisible(false);
                else
                    Menu.SetVisible(true);
                _grid.UpdateLayout();
            }
            VScrollBar.Slider.Handler.SetDrawable(true);
            total_invisible_size -= visible_area;
            v_size = total_invisible_size;

            if (total_invisible_size > 0)
            {
                float size_handler = (float)(visible_area)
                    / (float)total * 100.0f;
                size_handler = (float)VScrollBar.Slider.GetHeight() / 100.0f * size_handler;
                //size of handler
                VScrollBar.Slider.Handler.SetHeight((int)size_handler);
            }
            //step of slider
            float step_count = total_invisible_size / _area.GetStep();
            VScrollBar.Slider.SetStep((VScrollBar.Slider.GetMaxValue() - VScrollBar.Slider.GetMinValue()) / step_count);
            VScrollBar.Slider.SetCurrentValue((100.0f / total_invisible_size) * _area.GetVScrollOffset());
        }
        private void UpdateHorizontalSlider()//horizontal slider
        {
            int max_size = 0;
            int visible_area = _area.GetWidth() - _area.GetPadding().Left - _area.GetPadding().Right;
            if (visible_area < 0)
                visible_area = 0;
            foreach (var item in _area.GetItems())
            {
                if (!item.IsVisible())
                    continue;
                if (max_size < item.GetWidth() + item.GetMargin().Left + item.GetMargin().Right)
                    max_size = item.GetWidth() + item.GetMargin().Left + item.GetMargin().Right;
            }
            if (max_size <= visible_area)
            {
                HScrollBar.Slider.Handler.SetDrawable(false);
                HScrollBar.Slider.SetStep(HScrollBar.Slider.GetMaxValue());
                h_size = 0;
                HScrollBar.Slider.SetCurrentValue(0);
                if (GetHScrollBarVisible() == ScrollBarVisibility.AsNeeded)
                {
                    HScrollBar.SetDrawable(false);
                    Menu.SetVisible(false);
                    _grid.UpdateLayout();
                }
                return;
            }
            if (GetHScrollBarVisible() == ScrollBarVisibility.AsNeeded)
            {
                HScrollBar.SetDrawable(true);
                if (!VScrollBar.IsDrawable())
                    Menu.SetVisible(false);
                else
                    Menu.SetVisible(true);
                _grid.UpdateLayout();
            }
            HScrollBar.Slider.Handler.SetDrawable(true);
            int total_invisible_size = max_size - visible_area;
            h_size = total_invisible_size;

            if (total_invisible_size > 0)
            {
                float size_handler = (float)(visible_area) / (float)max_size * 100.0f;
                size_handler = (float)HScrollBar.Slider.GetWidth() / 100.0f * size_handler;
                //size of handler
                HScrollBar.Slider.Handler.SetWidth((int)size_handler);
            }
            //step of slider
            int step_count = (int)((float)total_invisible_size / (float)_area.GetStep());
            if (step_count == 0) HScrollBar.Slider.SetStep(HScrollBar.Slider.GetMaxValue());
            else
                HScrollBar.Slider.SetStep((HScrollBar.Slider.GetMaxValue() - HScrollBar.Slider.GetMinValue()) / step_count);

            float f = (100.0f / (float)total_invisible_size) * (float)_area.GetHScrollOffset();
            HScrollBar.Slider.SetCurrentValue(f);
        }

        /// <summary>
        /// Set width of the ListBox
        /// </summary>
        public override void SetWidth(int width)
        {
            base.SetWidth(width);
            UpdateHorizontalSlider();
            HScrollBar.Slider.UpdateHandler();
        }

        /// <summary>
        /// Set height of the ListBox
        /// </summary>
        public override void SetHeight(int height)
        {
            base.SetHeight(height);
            UpdateVerticalSlider();
            VScrollBar.Slider.UpdateHandler();
        }

        /// <summary>
        /// Add item to the ListBox
        /// </summary>
        public override void AddItem(IBaseItem item)
        {
            _area.AddItem(item);
            UpdateElements();
        }

        /// <summary>
        /// Insert item to the ListBox by index
        /// </summary>
        public override void InsertItem(IBaseItem item, Int32 index)
        {
            _area.InsertItem(item, index);
            UpdateElements();
        }

        /// <summary>
        /// Remove item from the ListBox
        /// </summary>
        public override bool RemoveItem(IBaseItem item)
        {
            List<IBaseItem> list = new List<IBaseItem>(GetItems());
            if (list.Contains(item))
            {
                return base.RemoveItem(item);
            }
            bool b = _area.RemoveItem(item);
            UpdateElements();
            _area.SetFocus();
            return b;
        }
        public override void Clear()
        {
            _area.Clear();
            // UpdateElements();
        }

        /// <summary>
        /// Update states of the all ListBox inner items
        /// </summary>
        public void UpdateElements()
        {
            UpdateVerticalSlider();
            VScrollBar.Slider.UpdateHandler();
            UpdateHorizontalSlider();
            HScrollBar.Slider.UpdateHandler();
        }

        /// <summary>
        /// Initialization and adding of all elements in the ListBox
        /// </summary>
        public override void InitElements()
        {
            //Adding
            base.AddItem(_grid);

            _grid.InsertItem(_area, 0, 0);
            _grid.InsertItem(VScrollBar, 0, 1);
            _grid.InsertItem(HScrollBar, 1, 0);
            _grid.InsertItem(Menu, 1, 1);

            //Events Connections
            EventScrollUp += (sender, args) =>
            {
                if (args.Mods == 0)
                    VScrollBar.EventScrollUp.Invoke(sender, args);
            };

            EventScrollDown += (sender, args) =>
            {
                if (args.Mods == 0)
                    VScrollBar.EventScrollDown.Invoke(sender, args);
            };
            _area.ItemListChanged += () => { UpdateElements(); };

            VScrollBar.Slider.EventValueChanged += (sender) => { UpdateVListArea(); };
            HScrollBar.Slider.EventValueChanged += (sender) => { UpdateHListArea(); };

            // create menu
            if (!_is_menu_disabled)
            {
                _menu = new ContextMenu(GetHandler());
                _menu.SetBackground(60, 60, 60);
                _menu.SetPassEvents(false);

                MenuItem go_up = new MenuItem("Go up");
                go_up.SetForeground(Color.FromArgb(210, 210, 210));
                go_up.EventMouseClick += ((sender, args) =>
                {
                    VScrollBar.Slider.SetCurrentValue(VScrollBar.Slider.GetMinValue());
                });

                MenuItem go_down = new MenuItem("Go down");
                go_down.SetForeground(Color.FromArgb(210, 210, 210));
                go_down.EventMouseClick += ((sender, args) =>
                {
                    VScrollBar.Slider.SetCurrentValue(VScrollBar.Slider.GetMaxValue());
                });

                MenuItem go_up_left = new MenuItem("Go up and left");
                go_up_left.SetForeground(Color.FromArgb(210, 210, 210));
                go_up_left.EventMouseClick += ((sender, args) =>
                {
                    VScrollBar.Slider.SetCurrentValue(VScrollBar.Slider.GetMinValue());
                    HScrollBar.Slider.SetCurrentValue(HScrollBar.Slider.GetMinValue());
                });

                MenuItem go_down_right = new MenuItem("Go down and right");
                go_down_right.SetForeground(Color.FromArgb(210, 210, 210));
                go_down_right.EventMouseClick += ((sender, args) =>
                {
                    VScrollBar.Slider.SetCurrentValue(VScrollBar.Slider.GetMaxValue());
                    HScrollBar.Slider.SetCurrentValue(HScrollBar.Slider.GetMaxValue());
                });
                _menu.AddItems(go_up_left, go_down_right, go_up, go_down);
                Menu.EventMouseClick += (sender, args) =>
                {
                    if (!_is_menu_disabled)
                        _menu.Show(sender, args);
                };
                _menu.ActiveButton = MouseButton.ButtonLeft;
                _menu.SetShadow(10, 0, 0, Color.Black);
            }
        }

        /// <returns> list of all ListBox items </returns>
        public List<IBaseItem> GetListContent()
        {
            List<IBaseItem> result = new List<IBaseItem>();
            List<IBaseItem> list = new List<IBaseItem>(_area.GetItems());

            foreach (IBaseItem item in list)
            {
                SelectionItem tmp = item as SelectionItem;
                // if (tmp != null)
                result.Add(tmp.GetContent());
                // Console.WriteLine(item.GetItemName() + " " + GetParent().GetItemName() + " " + GetHandler().GetWindowName());
            }
            return result;
        }

        /// <summary>
        /// Set list of items
        /// </summary>
        public virtual void SetListContent(List<IBaseItem> content)
        {
            _area.RemoveAllItems();
            foreach (IBaseItem item in content)
                AddItem(item);
        }

        public SelectionItem GetWrapper(IBaseItem item)
        {
            return GetArea()._mapContent[item];
        }

        /// <returns> selection item </returns>
        public IBaseItem GetSelectedItem()
        {
            return _area.GetSelectedItem();
        }

        /// <summary>
        /// Set style of the ListBox
        /// </summary>
        //style
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);

            Style inner_style = style.GetInnerStyle("vscrollbar");
            if (inner_style != null)
            {
                VScrollBar.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("hscrollbar");
            if (inner_style != null)
            {
                HScrollBar.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("menu");
            if (inner_style != null)
            {
                Menu.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("area");
            if (inner_style != null)
            {
                _area.SetStyle(inner_style);
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/ListBox.java
package com.spvessel;

import com.spvessel.Common.DefaultsService;
import com.spvessel.Core.InterfaceBaseItem;
import com.spvessel.Core.InterfaceCommonMethod;
import com.spvessel.Core.InterfaceCommonMethodState;
import com.spvessel.Core.InterfaceItem;
import com.spvessel.Core.InterfaceMouseMethodState;
import com.spvessel.Decorations.Style;
import com.spvessel.Flags.MouseButton;
import com.spvessel.Flags.ScrollBarVisibility;

import java.awt.Color;
import java.util.LinkedList;
import java.util.List;

public class ListBox extends Prototype {
    private static int count = 0;

    /**
     * Selection area of the ListBox
     */
    public int getSelection() {
        return _area.getSelection();
    }

    /**
     * Set selected item of the ListBox by index
     */
    public void setSelection(int index) {
        _area.setSelection(index);
    }

    /**
     * Unselect all items
     */
    public void unselect() {
        _area.unselect();
    }

    /**
     * Is selection changes view of the item or not
     */
    public void setSelectionVisibility(boolean visibility) {
        _area.setSelectionVisibility(visibility);
    }
    public boolean getSelectionVisibility() {
        return _area.getSelectionVisibility();
    }

    private Grid _grid = new Grid(2, 2);
    private ListArea _area = new ListArea();

    /**
     * @return ListArea
     */
    public ListArea getArea() {
        return _area;
    }

    public BlankItem menu = new BlankItem();
    private boolean _is_menu_disabled = false;

    /**
     * Is ListBox menu disabled
     */
    public void disableMenu(boolean value) {
        _is_menu_disabled = value;
    }

    private ContextMenu _menu;
    public VerticalScrollBar vScrollBar = new VerticalScrollBar();
    public HorizontalScrollBar hScrollBar = new HorizontalScrollBar();
    private ScrollBarVisibility _v_scrollBarPolicy = ScrollBarVisibility.ALWAYS;

    /**
     * Is vertical scroll bar visible
     */
    public ScrollBarVisibility getVScrollBarVisible() {
        return _v_scrollBarPolicy;
    }
    public void setVScrollBarVisible(ScrollBarVisibility policy) {
        _v_scrollBarPolicy = policy;

        if (policy == ScrollBarVisibility.NEVER) {
            vScrollBar.setDrawable(false);
            menu.setVisible(false);
        } else if (policy == ScrollBarVisibility.AS_NEEDED) {
            vScrollBar.setDrawable(false);
            menu.setVisible(false);
        } else if (policy == ScrollBarVisibility.ALWAYS) {
            vScrollBar.setDrawable(true);
            if (!hScrollBar.isDrawable())
                menu.setVisible(false);
            else
                menu.setVisible(true);
        }

        _grid.updateLayout();
        updateVerticalSlider();
        vScrollBar.slider.updateHandler();
    }

    private ScrollBarVisibility _h_scrollBarPolicy = ScrollBarVisibility.ALWAYS;

    /**
     * Is horizontal scroll bar visible
     */
    public ScrollBarVisibility getHScrollBarVisible() {
        return _h_scrollBarPolicy;
    }
    public void setHScrollBarVisible(ScrollBarVisibility policy) {
        _h_scrollBarPolicy = policy;

        if (policy == ScrollBarVisibility.NEVER) {
            hScrollBar.setDrawable(false);
            menu.setVisible(false);
        } else if (policy == ScrollBarVisibility.AS_NEEDED) {
            hScrollBar.setDrawable(false);
            menu.setVisible(false);
        } else if (policy == ScrollBarVisibility.ALWAYS) {
            hScrollBar.setDrawable(true);
            if (!vScrollBar.isDrawable())
                menu.setVisible(false);
            else
                menu.setVisible(true);
        }

        _grid.updateLayout();
        updateHorizontalSlider();
        hScrollBar.slider.updateHandler();
    }

    /**
     * Constructs a ListBox
     */
    public ListBox() {
        setItemName("ListBox_" + count);
        count++;
        // setStyle(DefaultsService.getDefaultStyle("SpaceVIL.ListBox"));
        setStyle(DefaultsService.getDefaultStyle(ListBox.class));

        // VBar
        vScrollBar.setDrawable(true);
        vScrollBar.setItemName(getItemName() + "_vScrollBar");

        // HBar
        hScrollBar.setDrawable(true);
        hScrollBar.setItemName(getItemName() + "_hScrollBar");

        // Area
        _area.setItemName(getItemName() + "_area");

        menu.setItemName(getItemName() + "_menu");
    }

    private void updateListAreaAttributes(InterfaceItem sender) {
        updateVListArea();
        updateHListArea();
    }

    private long v_size = 0;
    private long h_size = 0;

    private void updateVListArea() {
        // vertical slider
        float v_value = vScrollBar.slider.getCurrentValue();
        int v_offset = (int) ((float) (v_size * v_value) / 100.0f);
        _area.setVScrollOffset(v_offset);
    }

    private void updateHListArea() {
        // horizontal slider
        float h_value = hScrollBar.slider.getCurrentValue();
        int h_offset = (int) ((float) (h_size * h_value) / 100.0f);
        _area.setHScrollOffset(h_offset);
    }

    // vertical slider
    private void updateVerticalSlider() {
        int total_invisible_size = 0;
        int visible_area = _area.getHeight() - _area.getPadding().top - _area.getPadding().bottom;
        for (InterfaceBaseItem item : _area.getItems()) {
            if (item.equals(_area.getSubstrate()) || item.equals(_area.getHoverSubstrate()) || !item.isVisible())
                continue;
            total_invisible_size += (item.getHeight() + _area.getSpacing().vertical);
        }
        int total = total_invisible_size - _area.getSpacing().vertical;
        if (total_invisible_size <= visible_area) {
            vScrollBar.slider.handler.setDrawable(false);
            vScrollBar.slider.setStep(vScrollBar.slider.getMaxValue());
            v_size = 0;
            vScrollBar.slider.setCurrentValue(0);
            if (getVScrollBarVisible() == ScrollBarVisibility.AS_NEEDED) {
                vScrollBar.setDrawable(false);
                menu.setVisible(false);
                _grid.updateLayout();
            }
            return;
        }
        if (getVScrollBarVisible() == ScrollBarVisibility.AS_NEEDED) {
            vScrollBar.setDrawable(true);
            if (!hScrollBar.isDrawable())
                menu.setVisible(false);
            else
                menu.setVisible(true);
            _grid.updateLayout();
        }
        vScrollBar.slider.handler.setDrawable(true);
        total_invisible_size -= visible_area;
        v_size = total_invisible_size;

        if (total_invisible_size > 0) {
            float size_handler = (float) (visible_area) / (float) total * 100.0f;
            size_handler = (float) vScrollBar.slider.getHeight() / 100.0f * size_handler;
            // size of handler
            vScrollBar.slider.handler.setHeight((int) size_handler);
        }
        // step of slider
        float step_count = total_invisible_size / _area.getStep();
        vScrollBar.slider.setStep((vScrollBar.slider.getMaxValue() - vScrollBar.slider.getMinValue()) / step_count);
        vScrollBar.slider.setCurrentValue((100.0f / total_invisible_size) * _area.getVScrollOffset());
    }

    // horizontal slider
    private void updateHorizontalSlider() {
        int max_size = 0;
        int visible_area = _area.getWidth() - _area.getPadding().left - _area.getPadding().right;
        for (InterfaceBaseItem item : _area.getItems()) {
            if (item.equals(_area.getSubstrate()) || item.equals(_area.getHoverSubstrate()) || !item.isVisible())
                continue;
            if (max_size < item.getWidth() + item.getMargin().left + item.getMargin().right)
                max_size = item.getWidth() + item.getMargin().left + item.getMargin().right;
        }
        if (max_size <= visible_area) {
            // hScrollBar.slider.handler.setWidth(/* hScrollBar.slider.getWidth() */0);
            hScrollBar.slider.handler.setDrawable(false);
            hScrollBar.slider.setStep(hScrollBar.slider.getMaxValue());
            h_size = 0;
            hScrollBar.slider.setCurrentValue(0);
            if (getHScrollBarVisible() == ScrollBarVisibility.AS_NEEDED) {
                hScrollBar.setDrawable(false);
                menu.setVisible(false);
                _grid.updateLayout();
            }
            return;
        }
        if (getHScrollBarVisible() == ScrollBarVisibility.AS_NEEDED) {
            hScrollBar.setDrawable(true);
            if (!vScrollBar.isDrawable())
                menu.setVisible(false);
            else
                menu.setVisible(true);
            _grid.updateLayout();
        }
        hScrollBar.slider.handler.setDrawable(true);
        int total_invisible_size = max_size - visible_area;
        h_size = total_invisible_size;

        if (total_invisible_size > 0) {
            float size_handler = (float) (visible_area) / (float) max_size * 100.0f;
            size_handler = (float) hScrollBar.slider.getWidth() / 100.0f * size_handler;
            // size of handler
            hScrollBar.slider.handler.setWidth((int) size_handler);
        }
        // step of slider
        int step_count = (int) ((float) total_invisible_size / (float) _area.getStep());
        if (step_count == 0)
            hScrollBar.slider.setStep(hScrollBar.slider.getMaxValue());
        else
            hScrollBar.slider.setStep((hScrollBar.slider.getMaxValue() - hScrollBar.slider.getMinValue()) / step_count);

        float f = (100.0f / (float) total_invisible_size) * (float) _area.getHScrollOffset();
        hScrollBar.slider.setCurrentValue(f);
    }

    /**
     * Set width of the ListBox
     */
    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        updateHorizontalSlider();
        hScrollBar.slider.updateHandler();
    }

    /**
     * Set height of the ListBox
     */
    @Override
    public void setHeight(int height) {
        super.setHeight(height);
        updateVerticalSlider();
        vScrollBar.slider.updateHandler();
    }

    /**
     * Add item to the ListBox
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        _area.addItem(item);
        updateElements();
    }

    /**
     * Insert item to the ListBox by index
     */
    @Override
    public void insertItem(InterfaceBaseItem item, int index) {
        _area.insertItem(item, index);
        updateElements();
    }

    /**
     * Remove item from the ListBox
     */
    @Override
    public void removeItem(InterfaceBaseItem item) {
        _area.removeItem(item);
        updateElements();
    }

    /**
     * Update states of the all ListBox inner items
     */
    public void updateElements() {
        updateVerticalSlider();
        vScrollBar.slider.updateHandler();
        updateHorizontalSlider();
        hScrollBar.slider.updateHandler();
    }

    /**
     * Initialization and adding of all elements in the ListBox
     */
    @Override
    public void initElements() {
        // Adding
        super.addItem(_grid);
        _grid.insertItem(_area, 0, 0);
        _grid.insertItem(vScrollBar, 0, 1);
        _grid.insertItem(hScrollBar, 1, 0);
        _grid.insertItem(menu, 1, 1);

        // Events Connections
        _area.itemListChanged.add(this::updateElements);

        //InterfaceMouseMethodState scroll_up = (sender, args) -> vScrollBar.eventScrollUp.execute(sender, args);
        eventScrollUp.add(vScrollBar.eventScrollUp::execute);
        //InterfaceMouseMethodState scroll_down = (sender, args) -> vScrollBar.eventScrollDown.execute(sender, args);
        eventScrollDown.add(vScrollBar.eventScrollDown::execute);

        InterfaceCommonMethodState v_changed = (sender) -> updateVListArea();
        vScrollBar.slider.eventValueChanged.add(v_changed);
        InterfaceCommonMethodState h_changed = (sender) -> updateHListArea();
        hScrollBar.slider.eventValueChanged.add(h_changed);

        // create menu
        _menu = new ContextMenu(getHandler());
        _menu.setBackground(60, 60, 60);
        _menu.setPassEvents(false);

        MenuItem go_up = new MenuItem("Go up");
        go_up.setForeground(new Color(210, 210, 210));
        go_up.eventMouseClick.add((sender, args) -> {
            vScrollBar.slider.setCurrentValue(vScrollBar.slider.getMinValue());
        });

        MenuItem go_down = new MenuItem("Go down");
        go_down.setForeground(new Color(210, 210, 210));
        go_down.eventMouseClick.add((sender, args) -> {
            vScrollBar.slider.setCurrentValue(vScrollBar.slider.getMaxValue());
        });

        MenuItem go_up_left = new MenuItem("Go up and left");
        go_up_left.setForeground(new Color(210, 210, 210));
        go_up_left.eventMouseClick.add((sender, args) -> {
            hScrollBar.slider.setCurrentValue(hScrollBar.slider.getMinValue());
            vScrollBar.slider.setCurrentValue(vScrollBar.slider.getMinValue());
        });

        MenuItem go_down_right = new MenuItem("Go down and right");
        go_down_right.setForeground(new Color(210, 210, 210));
        go_down_right.eventMouseClick.add((sender, args) -> {
            hScrollBar.slider.setCurrentValue(hScrollBar.slider.getMaxValue());
            vScrollBar.slider.setCurrentValue(vScrollBar.slider.getMaxValue());
        });
        _menu.addItems(go_up_left, go_down_right, go_up, go_down);
        menu.eventMouseClick.add((sender, args) -> {
            if (!_is_menu_disabled)
                _menu.show(sender, args);
        });
        _menu.activeButton = MouseButton.BUTTON_LEFT;
        _menu.setShadow(10, 0, 0, Color.black);
    }

    /**
     * @return list of all ListBox items
     */
    public List<InterfaceBaseItem> getListContent() {
        List<InterfaceBaseItem> result = new LinkedList<>();
        for (InterfaceBaseItem item : _area.getItems()) {
            if (item.equals(_area.getSubstrate()) || item.equals(_area.getHoverSubstrate()))
                continue;
            result.add(item);
        }
        return result;
    }

    /**
     * Set list of items
     */
    public void setListContent(List<InterfaceBaseItem> content) {
        content.add(0, _area.getSubstrate());
        content.add(1, _area.getHoverSubstrate());
        _area.setContent(content);
    }

    /**
     * @return selection item
     */
    public InterfaceBaseItem getSelectionItem() {
        //List<InterfaceBaseItem> result = new LinkedList<>();
        return _area.getSelectionItem();
    }

    /**
     * Set style of the ListBox
     */
    // style
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        Style inner_style = style.getInnerStyle("vscrollbar");
        if (inner_style != null) {
            vScrollBar.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("hscrollbar");
        if (inner_style != null) {
            hScrollBar.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("menu");
        if (inner_style != null) {
            menu.setStyle(inner_style);
        }
    }
}